Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/TargetController.cs

```diff
@@ -0,0 +1,143 @@
+﻿using UnityEngine;
+using Random = UnityEngine.Random;
+using Unity.MLAgents;
+using UnityEngine.Events;
+
+namespace Unity.MLAgentsExamples
+{
+    /// <summary>
+    /// Utility class to allow target placement and collision detection with an agent
+    /// Add this script to the target you want the agent to touch.
+    /// Callbacks will be triggered any time the target is touched with a collider tagged as 'tagToDetect'
+    /// </summary>
+    public class TargetController : MonoBehaviour
+    {
+        
+        [Header("Collider Tag To Detect")]
+        public string tagToDetect = "agent"; //collider tag to detect 
+        
+        [Header("Target Placement")]
+        public float spawnRadius; //The radius in which a target can be randomly spawned.
+        public bool respawnIfTouched; //Should the target respawn to a different position when touched
+        
+        [Header("Target Fell Protection")]
+        public bool respawnIfFallsOffPlatform = true; //If the target falls off the platform, reset the position.
+        public float fallDistance = 5; //distance below the starting height that will trigger a respawn 
+
+
+        private Vector3 m_startingPos; //the starting position of the target
+        private Agent m_agentTouching; //the agent currently touching the target
+
+        [System.Serializable]
+        public class TriggerEvent : UnityEvent<Collider>
+        {
+        }
+
+        [Header("Trigger Callbacks")] public bool triggerIsTouching;
+        public TriggerEvent onTriggerEnterEvent = new TriggerEvent();
+        public TriggerEvent onTriggerStayEvent = new TriggerEvent();
+        public TriggerEvent onTriggerExitEvent = new TriggerEvent();
+
+        [System.Serializable]
+        public class CollisionEvent : UnityEvent<Collision>
+        {
+        }
+
+        [Header("Collision Callbacks")] public bool colliderIsTouching;
+        public CollisionEvent onCollisionEnterEvent = new CollisionEvent();
+        public CollisionEvent onCollisionStayEvent = new CollisionEvent();
```

Comment:
Why do we expose those ? Do we expect users to subscribe to these events ?

Reply:
yeah they show as events in the editor so the user won't have to subscribe/unsubscribe via code. just drag in the object you want to send the callback to and choose the method.

Reply:
Is the user expected to use / modify these callbacks themselves? There is no tutorial on how to do so, and the use case seems pretty limited to me.

Reply:
The user will use the standard event editor workflow. Choose the event you want. Drag in the object that needs to receive the callback and choose the method you want to call. We could separate out the collision/trigger events into it's own script. That's how i do it in other projects. 
<img width="517" alt="Screen Shot 2020-07-02 at 12 19 19 PM" src="https://user-images.githubusercontent.com/19273127/86401079-7ee36180-bc5e-11ea-9853-529672bfeae7.png">